helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -795,6 +798,55 @@ private RebalanceType getRebalanceType(ResourceControllerDataProvider cache,
     }
   }
 
+  private RebalanceType getRebalanceTypePerMessage(Map<String, Integer> requiredStates, Message message,
+      Map<String, String> derivedCurrentStates) {
+    Map<String, Integer> requiredStatesSnapshot = new HashMap<>(requiredStates);
+    // Looping existing current states to see whether current states fulfilled all the required states.
+    for (String state : derivedCurrentStates.values()) {
+      if (requiredStatesSnapshot.containsKey(state)) {
+        requiredStatesSnapshot.put(state, requiredStatesSnapshot.get(state) - 1);
+        if (requiredStatesSnapshot.get(state) == 0) {
+          requiredStatesSnapshot.remove(state);
+        }
+      }
+    }
```

Comment:
This calculation would be done multiple times if a partition has more than one message. Should we just pre-calculate the requiredStatesSnapshot (which is basically the gap between the derived current state and required states) and then use the snapshot? Obviously, we need to update the snapshot once a message is confirmed. But that calculation would be much faster than loop these 2 types of maps every time.

Reply:
No. We cannot, because the required state can be update and will be more complicated just update the required state. For example, in the state of: required state 1 LEADER, 1 FOLLOWER
1: FOLLOWER
2: OFFLINE

required state : 1 LEADER
with first message FOLLOWER -> LEADER applied.
required state : 1 FOLLOWER

So with this message apply, you need to give back the FOLLOWER. It will be complicated to determine when we should give the state back to required and when is not. I am not saying we cannot do that but just make the logic more complicated.

Reply:
I think the time complexities of the 2 methods are very different. O(n^2): (n + n-1 + n-2...) vs, O(n). I get the point that logic would be a little bit complicated, but I would still prefer updating the state map on the fly instead repeat the same calculation.

Reply:
Synced offline. This could involve more complicated logic such as return state back over the size of the what it required. Then the following computation will not be accurate. Will keep this as it and see whether it could cause more delay in computation.